README
README for Perl extension WWW::Curl

The Perl module WWW::Curl provides an interface to the cURL library "libcurl".
See http://curl.haxx.se/ for more information on cURL and libcurl.

This module requires libcurl and the corresponding header files to be
installed. 

For example on Ubuntu installing libcurl4-openssl-dev will satisfy this
dependency.

You then may install this module via the usual way
(for installation on Windows please see README.Win32):

    perl Makefile.PL
    make
    make test
    make install

The Makefile.PL script will try to use the 'curl-config' command to find
the correct headers and libcurl path. If curl.h can't be found,
either through curl-config or the internal list of guesses, you can specify
an explicit path when building the Makefile:

    perl Makefile.PL /home/foo/curl/include

If you want to specify which curl-config to use instead of the first in path 
set the environment variable CURL_CONFIG to point at the curl-config to use.

Minimum version requirements:

This module, starting from version 4.18 requires at least Perl 5.8.1 and libcurl 7.10.8.
These releases are more than 5 years old and they have multiple security vulnerabilities,
it is advised that you upgrade as soon as possible to a more recent, secure version.
Anything older than these versions of Perl and libcurl respectively is not supported.

Windows specific fixes and patches are welcome as testing is mainly focused on Linux.

The module provides the same functionality, except as noted in the documentation, 
as libcurl provides to C programs. Please refer to the documentation of libcurl for the
general interface description and the WWW::Curl POD for the differences.

This software is distributed AS IS, WITHOUT WARRANTY OF ANY KIND, either
express or implied.

History:

The author of the original relase of this software is Georg Horn <horn at koblenz-net.de>

Parts of the callback support were added Forrest Cahoon
<forrest.cahoon at merrillcorp.com>

More callback support, many tests, additional documentation, and Makefile
features have been added by Cris Bailiff <c.bailiff+curl at devsecure.com>

Curl multi support has been added by Sebastian Riedel <sri at oook.de>

The current maintainers are Cris Bailiff <c.bailiff+curl at devsecure.com> and
Bálint Szilakszi <szbalint at cpan.org>

The latest version is available on CPAN - https://metacpan.org/pod/WWW::Curl .

The source code repository is available at https://github.com/szbalint/WWW--Curl .
 
Copyright (C) 2000-2005, 2008-2014 Daniel Stenberg, Cris Bailiff, Bálint Szilakszi, et al.  
You may opt to use, copy, modify, merge, publish, distribute and/or sell
copies of the Software, and permit persons to whom the Software is
furnished to do so, under the terms of the MIT license.


Changes
Revision history for the Perl binding of libcurl, WWW::Curl.

    - Test if Curl object can be sent to callback as PROGRESSDTA [Alberto Simões];
    - Test SPEED_DOWNLOAD and CONTENT_TYPE properties [Alberto Simões];

4.17 Fri Feb 21 2014: - Balint Szilakszi <szbalint at cpan.org>

    - Fixing build process for old libcurl versions without CURLOPT_RESOLVE.
    - License is now MIT only.

4.16 Thu Feb 20 2014: - Balint Szilakszi <szbalint at cpan.org>

    - Support for CURLOPT_RESOLVE (an slist option) [Theo Schlossnagle]
    - Fixing t/19multi.t test failures when using a threaded resolver for libcurl.
    - Improved constant parsing when using ISO-compliant CPP. [tsibley]

4.15 Sun Nov 28 2010: - Balint Szilakszi <szbalint at cpan.org>

    - Refactored constant handling and added thorough testing for it.
    - Fixed CURLOPT_PRIVATE, it is now a string and can be set/get accordingly.

4.14 Sun Oct 24 2010: - Balint Szilakszi <szbalint at cpan.org>

    - Scalar references can now be used to receive body/header data [gfx].
    - Speed optimizations for threaded perl. [gfx, szbalint].
    - Added a more generic libcurl constant detection.
    - Added the pushopt method for appending strings to array options.
    - Documentation improvements.

4.13 Wed Sep 01 2010: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed WWW::Curl::Form (again, formadd and formaddfile working now).
    - Made constant constant handling more robust and added tests [Fuji, Goro].
    - Modernized *.pm and AUTOLOAD now throws an error on unknown method calls [Fuji, Goro].
    - Fixed code depending on CURLINFO_SLIST to be optional [Fuji, Goro].

4.12 Sun Jun 04 2010: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed WWW::Curl::Form (patch contributed by Michael Ing).
    - Added CURLPROXY_* constants (patch by claes).
    - Added the const_string method to provide access to constants in a non-bareword way.
    - Fixed t/01basic.t test failure on older libcurl versions.

4.11 Fri Dec 18 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed t/19multi.t for libcurl versions compiled with asyncronous dns resolution.

4.10 Fri Dec 18 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Added support for CURLINFO_SLIST in getinfo (patch by claes).
    - Merging documentation fixes upstream from the FreeBSD port (thanks Peter).
    - Added support for curl_multi_fdset.
    
4.09 Thu Jul 09 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Fixing broken version check.

4.08 Tue Jul 07 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed a memory leak in setopt.
    - Added a check to Makefile.PL for the minimum libcurl version.
    - Mentioned WWW::Curl hosting on github.
    - Upgraded bundled Module::Install to 0.91.

4.07 Sun May 31 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed >32bit integer option passing to libcurl on 32bit systems.
      (Thanks to Peter Heuchert for the report and fix suggestion!)
    - The CURL_CONFIG environment variable can now be used to specify
      which curl-config to use (contributed by claes).
    - Fixed segfault when a string option with setopt was set to undef
      (contributed by claes).
    - Fixed incomplete cleanup routine at destruction time
      (contributed by claes).
    - Readded Easy.pm and Share.pm stubs so that they are indexed by
      CPAN, thus avoiding complications with outdated versions appearing.

4.06 Sun Apr 05 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed a setopt issue that could cause a setopt string to get
      corrupted with pre 7.17.0 libcurl versions.
      (Thanks to Andreas Marienborg for the excellent bugreport and testcase!)
    - Made the Makefile.PL curl.h parsing slightly more generic to detect a
      wider selection of constants (fixes debian bugs #519841 and #518942).
      This makes it possible to use CURLE_* and CURL_NETRC_* constants.
    - Added a clearly invalid protocol to t/18twinhandles.t (fixes RT #41425).

4.05 Sat Jul 06 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Corrected the duphandle mechanism.
    - Fixed a typo (fixes RT #37253).
    - Added better Windows compatibility detection and separate
      installation instructions (thanks to Shiyu Tang).

4.04 Sat Jun 21 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Test::More and use_ok at begin time doesn't mesh with no plan.

4.03 Sat Jun 21 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Removing option to pass destination files as a typeglob as
      it is a backwards pointing magic and caused bugs.
      Use a filehandle or reference instead.
    - Fixed tests for 5.6, raised minimum Perl version to 5.6.1 due to tests.
    - Fixed a duphandle bug.

4.02 Thu Jun 19 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Adding forgotten backwards compatibility with older Perl versions

4.01 Wed Jun 18 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Reimplemented callback handling to fix several I/O issues (fixes RT #35491)
    - Documentation showing example WWW::Curl::Multi usage is corrected (fixes RT #35520)
    - Should now compile on HP-UX and OSX (fixes RT #36058 and #35349)
    - POD validity fix (fixes RT #15850)
    - Added several new tests
    - Upgraded to Module::Install 0.75 as it now exits correctly on build time
      with NA if the required libcurl isn't available.

4.00 Wed Mar 05 2008: - Balint Szilakszi <szbalint at cpan.org>
    - Rewritten documentation.
    - New test suite.
    - New build system: using Module::Install instead of ExtUtils::Makemaker.
    - New minimal requirements, Perl 5.6 and libcurl 7.10.8.
    - New WWW::Curl::Multi interface.
    - Removed WWW::Curl::easy.
    - Removed compatibility code for pre 7.10.8 libcurl.
    - Added lots of small improvements to setopt and friends.
    - Bugfixes, mainly for memory leaks.
    - SIGNATURE.

3.12 Wed Feb 27 2008: - Cris Bailiff <c.bailiff+curl at devsecure.com>

lib/WWW/Curl.pm
package WWW::Curl;

use strict;
use warnings;
use XSLoader;

our $VERSION = '4.17';
XSLoader::load(__PACKAGE__, $VERSION);

END {
    _global_cleanup();
}

1;

__END__

=head1 NAME

WWW::Curl - Perl extension interface for libcurl

=head1 SYNOPSIS

    use WWW::Curl;
    print $WWW::Curl::VERSION;


=head1 DESCRIPTION

WWW::Curl is a Perl extension interface for libcurl.

=head1 DOCUMENTATION

This module provides a Perl interface to libcurl. It is not intended to be a standalone module
and because of this, the main libcurl documentation should be consulted for API details at
L<http://curl.haxx.se>. The documentation you're reading right now only contains the Perl specific
details, some sample code and the differences between the C API and the Perl one.

=head1 WWW::Curl::Easy

The name might be confusing, it originates from libcurl. This is not an ::Easy module
in the sense normally used on CPAN.

Here is a small snippet of making a request with WWW::Curl::Easy.

    use strict;
    use warnings;
    use WWW::Curl::Easy;

    my $curl = WWW::Curl::Easy->new;
    
    $curl->setopt(CURLOPT_HEADER,1);
    $curl->setopt(CURLOPT_URL, 'http://example.com');

    # A filehandle, reference to a scalar or reference to a typeglob can be used here.
    my $response_body;
    $curl->setopt(CURLOPT_WRITEDATA,\$response_body);

    # Starts the actual request
    my $retcode = $curl->perform;

    # Looking at the results...
    if ($retcode == 0) {
        print("Transfer went ok\n");
        my $response_code = $curl->getinfo(CURLINFO_HTTP_CODE);
        # judge result and next action based on $response_code
        print("Received response: $response_body\n");
    } else {
        # Error code, type of error, error message
        print("An error happened: $retcode ".$curl->strerror($retcode)." ".$curl->errbuf."\n");
    }

See L<curl_easy_setopt(3)> for details of C<setopt()>.

=head1 WWW::Curl::Multi

    use strict;
    use warnings;
    use WWW::Curl::Easy;
    use WWW::Curl::Multi;

    my %easy;
    my $curl = WWW::Curl::Easy->new;
    my $curl_id = '13'; # This should be a handle unique id.
    $easy{$curl_id} = $curl;
    my $active_handles = 0;

    $curl->setopt(CURLOPT_PRIVATE,$curl_id);
    # do the usual configuration on the handle
    ...

    my $curlm = WWW::Curl::Multi->new;
    
    # Add some easy handles
    $curlm->add_handle($curl);
    $active_handles++;

    while ($active_handles) {
        my $active_transfers = $curlm->perform;
        if ($active_transfers != $active_handles) {
            while (my ($id,$return_value) = $curlm->info_read) {
                if ($id) {
                    $active_handles--;
                    my $actual_easy_handle = $easy{$id};
                    # do the usual result/error checking routine here
                    ...
                    # letting the curl handle get garbage collected, or we leak memory.
                    delete $easy{$id};
                }
            }
        }
    }

This interface is different than what the C API does. $curlm->perform is non-blocking and performs
requests in parallel. The method does a little work and then returns control, therefor it has to be called
periodically to get the job done. It's return value is the number of unfinished requests.

When the number of unfinished requests changes compared to the number of active handles, $curlm->info_read
should be checked for finished requests. It returns one handle and it's return value at a time, or an empty list
if there are no more finished requests. $curlm->info_read calls remove_handle on the given easy handle automatically,
internally. The easy handle will still remain available until it goes out of scope, this action just detaches it from
multi.

Please make sure that the easy handle does not get garbage collected until after the multi handle finishes processing it,
or bad things happen.

The multi handle does not need to be cleaned up, when it goes out of scope it calls the required cleanup methods
automatically.

It is possible to use $curlm->add_handle to add further requests to be processed after $curlm->perform has been called.
WWW::Curl::Multi doesn't care about the order. It is possible to process all requests for a multi handle and then add
a new batch of easy handles for processing.

=head1 WWW::Curl::Share

    use WWW::Curl::Share;
    my $curlsh = new WWW::Curl::Share;
    $curlsh->setopt(CURLSHOPT_SHARE, CURL_LOCK_DATA_COOKIE);
    $curlsh->setopt(CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
    $curl->setopt(CURLOPT_SHARE, $curlsh);
    $curlsh->setopt(CURLSHOPT_UNSHARE, CURL_LOCK_DATA_COOKIE);
    $curlsh->setopt(CURLSHOPT_UNSHARE, CURL_LOCK_DATA_DNS);

WWW::Curl::Share is an extension to WWW::Curl::Easy which makes it possible
to use a single cookies/dns cache for several Easy handles.

It's usable methods are:
    
    $curlsh = new WWW::Curl::Share
        This method constructs a new WWW::Curl::Share object.

    $curlsh->setopt(CURLSHOPT_SHARE, $value );
        Enables share for:
            CURL_LOCK_DATA_COOKIE    use single cookies database
            CURL_LOCK_DATA_DNS       use single DNS cache
    $curlsh->setopt(CURLSHOPT_UNSHARE, $value );
        Disable share for given $value (see CURLSHOPT_SHARE)

    $curlsh->strerror( ErrNo )
        This method returns a string describing the CURLSHcode error 
        code passed in the argument errornum.

This is how you enable sharing for a specific WWW::Curl::Easy handle:

    $curl->setopt(CURLOPT_SHARE, $curlsh)
        Attach share object to WWW::Curl::Easy instance

=head1 WWW::Curl::Form

    use WWW::Curl::Form;
    my $curlf = WWW::Curl::Form->new;
    $curlf->formaddfile($filename, 'attachment', "multipart/form-data");
    $curlf->formadd("FIELDNAME", "VALUE");

    $curl->setopt(CURLOPT_HTTPPOST, $curlf);

Its usable methods are:

    $curlf = new WWW::Curl::Form
        This method constructs a new WWW::Curl::Form object.

    $curlf->formadd(FIELDNAME, VALUE)
        This method adds a field with a given value, to the form that is being submitted.

    $curlf->formaddfile(FILENAME, DESCRIPTION, TYPE)
        This method will add a file to the form. The description is the name of the field
        that you form expects the data to be submitted in.

=head1 COMPATIBILITY

=over

=item curl_easy_setopt

Most of the options should work, however some might not. Please send reports, tests and patches to fix
those. 

=item curl_easy_escape

Not implemented. Since equivalent Perl code is easily produced, this method will only made
available for interface completeness, if ever.

=item curl_easy_init

Used only internally. The standard Perl way of initializing an object should be used,
 C<< my $curl = WWW::Curl::Easy->new; >>.

=item curl_easy_cleanup

Used only internally. Curl object cleanup happens when the handle goes out of scope.

=item curl_easy_duphandle

Should be working for most cases, however do not change the value of options which accept
a list/arrayref value on a duped handle, otherwise memory leaks or crashes will happen.
This behaviour will be fixed in the future.

=item curl_easy_pause

Not implemented.

=item curl_easy_reset

Not implemented.

=item curl_easy_unescape

Not implemented. Trivial Perl replacements are available.

=item curl_escape

Not implemented and won't be as this method is considered deprecated.

=item curl_formadd

Seems to be working.

=item curl_formaddfile

Seems to be working.

=item curl_formfree

Used internally. Not exposed through the public API, as this call has no relevance
to Perl code.

=item curl_free

Used internally. Not exposed through the public API, as this call has no relevance
to Perl code.

=item curl_getdate

Not implemented. This function is easily replaced by Perl code and as such, most likely
it won't be implemented.

=item curl_global_cleanup

Only used internally, not exposed through the public API.

=item curl_global_init

Only used internally, not exposed through the public API.

=item curl_global_init_mem

Not implemented.

=item curl_global_cleanup

Only used internally and called automatically upon exit.

=item curl_slist_append

Only used internally, not exposed through the public API.

=item curl_slist_free_all

Only used internally, not exposed through the public API.

=item curl_unescape

Not implemented and won't be, as this method is considered deprecated.

=item curl_version

Seems to work.

=item curl_version_info

Not yet implemented.

=item curl_multi_*

Most methods are either not exposed through the WWW::Curl::Multi API or they behave differently
than it's C counterpart. Please see the section about WWW::Curl::Multi above.

=item curl_multi_fdset

This method returns three arrayrefs: the read, write and exception fds libcurl knows about.
In the case of no file descriptors in the given set, an empty array is returned.

=back

=head1 NUANCES

=head2 Header output for redirects

It might be surprising that if C<CURLOPT_FOLLOWLOCATION> is set and header output was enabled, headers show up for all http responses.
The reasoning behind that and possible code adjustments are outlined here: L<https://rt.cpan.org/Ticket/Display.html?id=61569>.

=head2 CURLOPT_PRIVATE

Despite what the libcurl manual says, in Perl land, only string values are suitable for this option.

=head1 ADDITIONAL METHODS

=head2 On WWW::Curl::Easy objects

=over 

=item pushopt

Like C<setopt> but instead of overriding any previously set values it adds it 
to the end. Can be used with C<CURLOPT_HTTPHEADER>, C<CURLOPT_QUOTE> and 
C<CURLOPT_POSTQUOTE>.


=back

=head1 USAGE CASES

WWW::Curl is a thin binding on top of libcurl, to make using libcurl possible from Perl land.
Because of this, the module is less like Perl and more like C in coding style.

There is a new module, L<http://search.cpan.org/perldoc?WWW::Curl::Simple>,
which wraps this module into a more Perlish and userfriendly package.

The standard Perl WWW module, LWP should probably be used in most cases to work with HTTP or FTP from Perl.
However, there are some cases where LWP doesn't perform well. One is speed and the other is parallelism. 
WWW::Curl is much faster, uses much less CPU cycles and it's capable of non-blocking parallel requests.

In some cases, for example when building a web crawler, cpu usage and parallel downloads are
important considerations. It can be desirable to use WWW::Curl to do the heavy-lifting of
a large number of downloads and wrap the resulting data into a Perl-friendly structure by
HTTP::Response or use WWW::Curl::Simple to do that for you.

=head1 CHANGES

Version 4.01 - 4.07 adds several bugfixes and extends functionality coverage. See Changes file.

Version 4.00 added new documentation, the build system changed to Module::Install,
the test suite was rewritten to use Test::More, a new calling syntax for WWW::Curl::Multi
was added, memory leak and other bugfixes added, Perl 5.6 and libcurl 7.10.8 as minimum
requirements for this module were set.

Version 3.12 is a bugfix for a missing Share.pm.in file in the release.

Version 3.11 added WWW::Curl::Share.

Version 3.10 adds the WWW::Curl::Share interface by Anton Federov
and large file options after a contribution from Mark Hindley.

Version 3.02 adds some backwards compatibility for scripts still using
'WWW::Curl::easy' names.

Version 3.01 added some support for pre-multi versions of libcurl.

Version 3.00 adds WWW::Curl::Multi interface, and new module names
following perl conventions (WWW::Curl::Easy rather than WWW::Curl::easy),
by Sebastian Riedel <sri at cpan.org>.

Version 2.00 of WWW::Curl::easy is a renaming of the previous version
(named Curl::easy), to follow CPAN naming guidelines, by Cris Bailiff.

Versions 1.30, a (hopefully) threadable, object-oriented,
multiple-callback compatible version of Curl::easy was substantially
reworked from the previous Curl::easy release (1.21) by Cris Bailiff.

=head1 AUTHORS

Currently maintained by Cris Bailiff <c.bailiff+curl at devsecure.com> and
Balint Szilakszi <szbalint at cpan.org>.

Original Author Georg Horn <horn@koblenz-net.de>, with additional callback,
pod and test work by Cris Bailiff <c.bailiff+curl@devsecure.com> and
Forrest Cahoon <forrest.cahoon@merrillcorp.com>. Sebastian Riedel added ::Multi
and Anton Fedorov (datacompboy <at> mail.ru) added ::Share. Balint Szilakszi
repackaged the module into a more modern form.

=head1 COPYRIGHT

Copyright (C) 2000-2005,2008-2014 Daniel Stenberg, Cris Bailiff,
Sebastian Riedel, Balint Szilakszi et al.

You may opt to use, copy, modify, merge, publish, distribute and/or sell
copies of the Software, and permit persons to whom the Software is furnished
to do so, under the terms of the MIT license.

=head1 SEE ALSO

L<http://curl.haxx.se>

L<http://search.cpan.org/perldoc?WWW::Curl::Simple>

L<libcurl(3)>

The development source code is also available: L<http://github.com/szbalint/WWW--Curl/tree/master>
